sys/mips/mips/freebsd64_machdep.c

```diff
@@ -198,7 +200,224 @@ static void
+	    SIGISMEMBER(psp->ps_sigonstack, sig)) {
+		sp = (vm_offset_t)((__cheri_addr vaddr_t)td->td_sigstk.ss_sp +
+		    td->td_sigstk.ss_size);
+	} else {
+#ifdef CPU_CHERI
+		/*
+		 * Signals delivered when a CHERI sandbox is present must be
+		 * delivered on the alternative stack rather than a local one.
+		 * If an alternative stack isn't present, then terminate or
+		 * risk leaking capabilities (and control) to the sandbox (or
+		 * just crashing the sandbox).
+		 */
+		if (cheri_is_sandboxed) {
+			mtx_unlock(&psp->ps_mtx);
+			printf("pid %d, tid %d: signal in sandbox without "
+			    "alternative stack defined\n", td->td_proc->p_pid,
+			    td->td_tid);
+			sigexit(td, SIGILL);
+			/* NOTREACHED */
+		}
+#endif
+		sp = (vm_offset_t)regs->sp;
+	}
+#ifdef CPU_CHERI
+	cp2_len = sizeof(*cfp);
+	sp -= cp2_len;
+	sp &= ~(CHERICAP_SIZE - 1);
+	sf.sf_uc.uc_mcontext.mc_cp2state = sp;
+	sf.sf_uc.uc_mcontext.mc_cp2state_len = cp2_len;
+#endif
+	sp -= sizeof(struct sigframe);
+	sp = rounddown2(sp, STACK_ALIGN);
+	sfp = (struct sigframe64 *)sp;
+
+	/* Build the argument list for the signal handler. */
+	regs->a0 = sig;
+	regs->a2 = (register_t)(intptr_t)&sfp->sf_uc;
+	if (SIGISMEMBER(psp->ps_siginfo, sig)) {
+		/* Signal handler installed with SA_SIGINFO. */
+		regs->a1 = (register_t)(intptr_t)&sfp->sf_si;
+		/* sf.sf_ahu.sf_action = (__siginfohandler_t *)catcher; */
+
+		/* fill siginfo structure */
+		siginfo_to_siginfo64(&ksi->ksi_info, &sf.sf_si);
+		sf.sf_si.si_signo = sig;
+	} else {
+		/* Old FreeBSD-style arguments. */
+		regs->a1 = ksi->ksi_code;
+		regs->a3 = (__cheri_addr uintptr_t)ksi->ksi_addr;
+		/* sf.sf_ahu.sf_handler = catcher; */
+	}
+
+	mtx_unlock(&psp->ps_mtx);
+	PROC_UNLOCK(p);
+
+	/*
+	 * Copy the sigframe out to the user's stack.
+	 */
+#ifdef CPU_CHERI
+	cfp = malloc(sizeof(*cfp), M_TEMP, M_WAITOK);
+	cheri_trapframe_to_cheriframe(&td->td_pcb->pcb_regs, cfp);
+	if (copyoutcap(cfp,
+	    __USER_CAP((void *)(uintptr_t)sf.sf_uc.uc_mcontext.mc_cp2state,
+	    cp2_len), cp2_len) != 0) {
+		free(cfp, M_TEMP);
+		PROC_LOCK(p);
+		printf("pid %d, tid %d: could not copy out cheriframe\n",
+		    td->td_proc->p_pid, td->td_tid);
+		sigexit(td, SIGILL);
+		/* NOTREACHED */
+	}
+	free(cfp, M_TEMP);
+#endif
+	if (copyout(&sf, __USER_CAP_OBJ(sfp), sizeof(struct sigframe64)) != 0) {
+		/*
+		 * Something is wrong with the stack pointer.
+		 * ...Kill the process.
+		 */
+		PROC_LOCK(p);
+		printf("pid %d, tid %d: could not copy out sigframe\n",
+		    td->td_proc->p_pid, td->td_tid);
+		sigexit(td, SIGILL);
+		/* NOTREACHED */
+	}
+
+#ifdef CPU_CHERI
+	/*
+	 * Install CHERI signal-delivery register state for handler to run
+	 * in.  As we don't install this in the CHERI frame on the user stack,
+	 * it will be (genrally) be removed automatically on sigreturn().
+	 */
+	hybridabi_sendsig(td);
+#endif
+
+	regs->pc = (trapf_pc_t)catcher;
+	// FIXME: should this be an offset relative to PCC or an address?
```

Comment:
Now that regs->pc is $pcc, I guess this FIXME can go. 
$t9 should match $pc which is the offset of $pcc.

Reply:
Axed.